src/Microsoft.NetCore.Analyzers/Core/Runtime/UseToLowerInvariantOrToUpperInvariant.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using Analyzer.Utilities;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.NetCore.Analyzers.Runtime
+{
+    public abstract class UseToLowerInvariantOrToUpperInvariantAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA_IHaveNoIdeaHowTheseAreNumbered";
+
+        private static readonly LocalizableString s_localizableMessageAndTitle = new LocalizableResourceString(nameof(SystemRuntimeAnalyzersResources.UseToLowerInvariantOrToUpperInvariantTitle), SystemRuntimeAnalyzersResources.ResourceManager, typeof(SystemRuntimeAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(SystemRuntimeAnalyzersResources.UseToLowerInvariantOrToUpperInvariantDescription), SystemRuntimeAnalyzersResources.ResourceManager, typeof(SystemRuntimeAnalyzersResources));
+        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(RuleId,
+                                                                             s_localizableMessageAndTitle,
+                                                                             s_localizableMessageAndTitle,
+                                                                             DiagnosticCategory.Globalization,
+                                                                             DiagnosticHelpers.DefaultDiagnosticSeverity,
+                                                                             isEnabledByDefault: false,
+                                                                             description: s_localizableDescription);
+
+        internal const string ToLowerMethodName = "ToLower";
+        internal const string ToUpperMethodName = "ToUpper";
+
+        protected abstract Location GetMethodNameLocation(SyntaxNode invocationNode);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterOperationAction(operationContext =>
+            {
+                var operation = (IInvocationOperation)operationContext.Operation;
+                IMethodSymbol methodSymbol = operation.TargetMethod;
+
+                if (methodSymbol != null &&
```

Comment:
`IInvocationOperation.TargetMethod` is guaranteed to be non-null.

Reply:
Really? Cause I've seen it being checked in several places. Does that mean those checks should be removed too?

Reply:
Yes, likely that guarantee was not present in the IOperation API prototype prior to the eventual shipping API. Roslyn test suite runs the [TestOperationVisitor](http://source.roslyn.io/#Roslyn.Test.Utilities/Compilation/TestOperationVisitor.cs,19) on every Roslyn compiler unit test, which has the [following](http://source.roslyn.io/#Roslyn.Test.Utilities/Compilation/TestOperationVisitor.cs,458) assertion: `Assert.NotNull(operation.TargetMethod);`

Reply:
Understood, thanks for the detailed info. So how should we go about removing those redundant checks? I could offer PRs as I find them but I'm not sure if that's an efficient way of handling it. Any thoughts?

Reply:
You can submit a single PR to delete all null checks. If you are fully inspired to cleanup our entire IOperation usage null checks, you can also remove null checks for other APIs that are asserted to be non-null in the TestOperationVisitor. I'd let you make a call on the scope of your PR.

Reply:
@TKharaishvili If you want to do so, please do it as a separate PR from this one.